File: bindings/python/tokenizers/implementations/base_tokenizer.py
from tokenizers import Tokenizer, Encoding, AddedToken
from tokenizers.models import TokenizedSequence, TokenizedSequenceWithOffsets

from typing import List, Union, Tuple, Optional, Dict

Offsets = Tuple[int, int]


class BaseTokenizer:
    def __init__(self, tokenizer: Tokenizer, parameters=None):
        self._tokenizer = tokenizer
        self._parameters = parameters if parameters is not None else {}

    def __repr__(self):
        return "Tokenizer(vocabulary_size={}, {})".format(
            self._tokenizer.get_vocab_size(),
            ", ".join(k + "=" + str(v) for k, v in self._parameters.items()),
        )

    def num_special_tokens_to_add(self, is_pair: bool) -> int:
        """
        Return the number of special tokens that would be added for single/pair sentences.
        :param is_pair: Boolean indicating if the input would be a single sentence or a pair
        :return:
        """
        return self._tokenizer.num_special_tokens_to_add(is_pair)

    def get_vocab(self, with_added_tokens: bool = True) -> Dict[str, int]:
        """ Returns the vocabulary

        Args:
            with_added_tokens: boolean:
                Whether to include the added tokens in the vocabulary

        Returns:
            The vocabulary
        """
        return self._tokenizer.get_vocab(with_added_tokens=with_added_tokens)

    def get_vocab_size(self, with_added_tokens: bool = True) -> int:
        """ Return the size of vocabulary, with or without added tokens.

        Args:
            with_added_tokens: (`optional`) bool:
                Whether to count in added special tokens or not

        Returns:
            Size of vocabulary
        """
        return self._tokenizer.get_vocab_size(with_added_tokens=with_added_tokens)

    def enable_padding(
        self,
        direction: Optional[str] = "right",
        pad_id: Optional[int] = 0,
        pad_type_id: Optional[int] = 0,
        pad_token: Optional[str] = "[PAD]",
        max_length: Optional[int] = None,
    ):
        """ Change the padding strategy

        Args:
            direction: (`optional`) str:
                Can be one of: `right` or `left`

            pad_id: (`optional`) unsigned int:
                The indice to be used when padding

            pad_type_id: (`optional`) unsigned int:
                The type indice to be used when padding

            pad_token: (`optional`) str:
                The pad token to be used when padding

            max_length: (`optional`) unsigned int:
                If specified, the length at which to pad. If not specified
                we pad using the size of the longest sequence in a batch
        """
        return self._tokenizer.enable_padding(
            direction=direction,
            pad_id=pad_id,
            pad_type_id=pad_type_id,
            pad_token=pad_token,
            max_length=max_length,
        )

    def no_padding(self):
        """ Disable padding """
        return self._tokenizer.no_padding()

    def enable_truncation(
        self, max_length: int, stride: Optional[int] = 0, strategy: Optional[str] = "longest_first"
    ):
        """ Change the truncation options

        Args:
            max_length: unsigned int:
                The maximum length at which to truncate

            stride: (`optional`) unsigned int:
                The length of the previous first sequence to be included
                in the overflowing sequence

            strategy: (`optional) str:
                Can be one of `longest_first`, `only_first` or `only_second`
        """
        return self._tokenizer.enable_truncation(max_length, stride=stride, strategy=strategy)

    def no_truncation(self):
        """ Disable truncation """
        return self._tokenizer.no_truncation()

    def add_tokens(self, tokens: List[Union[str, AddedToken]]) -> int:
        """ Add the given tokens to the vocabulary

        Args:
            tokens: List[Union[str, AddedToken]]:
                A list of tokens to add to the vocabulary. Each token can either be
                a string, or an instance of AddedToken

        Returns:
            The number of tokens that were added to the vocabulary
        """
        return self._tokenizer.add_tokens(tokens)

    def add_special_tokens(self, special_tokens: List[Union[str, AddedToken]]) -> int:
        """ Add the given special tokens to the vocabulary, and treat them as special tokens.

        The special tokens will never be processed by the model, and will be
        removed while decoding.

        Args:
            tokens: List[Union[str, AddedToken]]:
                A list of special tokens to add to the vocabulary. Each token can either be
                a string, or an instance of AddedToken

        Returns:
            The number of tokens that were added to the vocabulary
        """
        return self._tokenizer.add_special_tokens(special_tokens)

    def normalize(self, sequence: str) -> str:
        """ Normalize the given sequence

        Args:
            sequence: str:
                The sequence to normalize

        Returns:
            The normalized string
        """
        return self._tokenizer.normalize(sequence)

    def encode_tokenized(
        self, sequence: Union[TokenizedSequence, TokenizedSequenceWithOffsets], type_id: int = 0
    ) -> Encoding:
        """ Encode the given sequence. Let us skip the Normalizer and PreTokenizer by providing
        already tokenized substrings.

        A sequence can either be:
            - `TokenizedSequence`: (`List[str]`)
            - `TokenizedSequenceWithOffsets: (`List[Tuple[str, Offsets]]`) where Offsets is
            a Tuple[int, int].

        If the Offsets are not provided, they will be automatically generated, making the hypothesis
        that all the tokens in the `TokenizedSequence` are contiguous in the original string.

        Args:
            sequence: Union[TokenizedSequence, TokenizedSequenceWithOffsets]
                Either a TokenizedSequence or a TokenizedSequenceWithOffsets

            type_id: int:
                The type id of the given sequence

        Returns:
            An Encoding
        """
        return self._tokenizer.model.encode(sequence, type_id)

    def encode_tokenized_batch(
        self,
        sequences: Union[List[TokenizedSequence], List[TokenizedSequenceWithOffsets]],
        type_id: int = 0,
    ) -> List[Encoding]:
        """ Encode the given batch of sequence. Let us skip the Normalizer and PreTokenizer by
        providing already tokenized substrings.

        A sequence can either be:
            - `TokenizedSequence`: (`List[str]`)
            - `TokenizedSequenceWithOffsets: (`List[Tuple[str, Offsets]]`) where Offsets is
            a Tuple[int, int].

        If the Offsets are not provided, they will be automatically generated, making the hypothesis
        that all the tokens in the `TokenizedSequence` are contiguous in the original string.

        Args:
            sequences: Union[List[TokenizedSequence], List[TokenizedSequenceWithOffsets]]
                A list of sequence. Each sequence is either a TokenizedSequence or a
                TokenizedSequenceWithOffsets

            type_id: int:
                The type if of the given sequence

        Returns:
            A list of Encoding
        """
        return self._tokenizer.model.encode_batch(sequences, type_id)

    def encode(
        self, sequence: str, pair: Optional[str] = None, add_special_tokens: bool = True
    ) -> Encoding:
        """ Encode the given sequence

        Args:
            sequence: str:
                The sequence to encode

            pair: (`optional`) Optional[str]:
                The optional pair sequence

            add_special_tokens: bool:
                Whether to add the special tokens while encoding.

        Returns:
            An Encoding
        """
        if sequence is None:
            raise ValueError("None input is not valid. Should be a string.")

        return self._tokenizer.encode(sequence, pair, add_special_tokens)

    def encode_batch(
        self, sequences: List[Union[str, Tuple[str, str]]], add_special_tokens: bool = True
    ) -> List[Encoding]:
        """ Encode the given sequences or pair of sequences

        Args:
            sequences: List[Union[str, Tuple[str, str]]]:
                A list of sequences or pair of sequences. The list can contain both
                at the same time.

            add_special_tokens: bool:
                Whether to add the special tokens while encoding.

        Returns:
            A list of Encoding
        """

        if sequences is None:
            raise ValueError(
                "None input is not valid. Should be a list of strings or a list of tuple of strings."
            )

        return self._tokenizer.encode_batch(sequences, add_special_tokens)

    def decode(self, ids: List[int], skip_special_tokens: Optional[bool] = True) -> str:
        """ Decode the given list of ids to a string sequence

        Args:
            ids: List[unsigned int]:
                A list of ids to be decoded

            skip_special_tokens: (`optional`) boolean:
                Whether to remove all the special tokens from the output string

        Returns:
            The decoded string
        """
        if ids is None:
            raise ValueError("None input is not valid. Should be a list of integers.")

        return self._tokenizer.decode(ids, skip_special_tokens=skip_special_tokens)

    def decode_batch(
        self, sequences: List[List[int]], skip_special_tokens: Optional[bool] = True
    ) -> str:
        """ Decode the list of sequences to a list of string sequences

        Args:
            sequences: List[List[unsigned int]]:
                A list of sequence of ids to be decoded

            skip_special_tokens: (`optional`) boolean:
                Whether to remove all the special tokens from the output strings

        Returns:
            A list of decoded strings
        """
        if sequences is None:
            raise ValueError("None input is not valid. Should be list of list of integers.")

        return self._tokenizer.decode_batch(sequences, skip_special_tokens=skip_special_tokens)

    def token_to_id(self, token: str) -> Optional[int]:
        """ Convert the given token to its corresponding id

        Args:
            token: str:
                The token to convert

        Returns:
            The corresponding id if it exists, None otherwise
        """
        return self._tokenizer.token_to_id(token)

    def id_to_token(self, id: int) -> Optional[str]:
        """ Convert the given token id to its corresponding string

        Args:
            token: id:
                The token id to convert

        Returns:
            The corresponding string if it exists, None otherwise
        """
        return self._tokenizer.id_to_token(id)

    def save(self, directory: str, name: Optional[str] = None):
        """ Save the current model to the given directory

        Args:
            directory: str:
                A path to the destination directory

            name: (Optional) str:
                The name of the tokenizer, to be used in the saved files
        """
        return self._tokenizer.model.save(directory, name=name)

    def post_process(
        self, encoding: Encoding, pair: Optional[Encoding] = None, add_special_tokens: bool = True
    ) -> Encoding:
        """ Apply all the post-processing steps to the given encodings.

        The various steps are:
            1. Truncate according to global params (provided to `enable_truncation`)
            2. Apply the PostProcessor
            3. Pad according to global params. (provided to `enable_padding`)

        Args:
            encoding: Encoding:
                The main Encoding to post process

            pair: Optional[Encoding]:
                An optional pair Encoding

            add_special_tokens: bool:
                Whether to add special tokens

        Returns:
            The resulting Encoding
        """
        return self._tokenizer.post_process(encoding, pair, add_special_tokens)

File: bindings/node/lib/implementations/tokenizers/base.tokenizer.ts
import { promisify } from "util";

import { PostProcessor } from "../../bindings/post-processors";
import {
  AddedToken,
  EncodeInput,
  EncodeOptions,
  InputSequence,
  PaddingConfiguration,
  PaddingOptions,
  Tokenizer,
  TruncationConfiguration,
  TruncationOptions,
} from "../../bindings/tokenizer";
import { Encoding } from "../encoding";

export type Token = string | AddedToken;

// eslint-disable-next-line @typescript-eslint/ban-types
export class BaseTokenizer<TConfig extends object> {
  private _truncation?: TruncationConfiguration;
  private _padding?: PaddingConfiguration;

  constructor(
    protected tokenizer: Tokenizer,
    /**
     * @since 0.4.0
     */
    readonly configuration: Readonly<TConfig>
  ) {}

  /**
   * Instantiate a new Tokenizer from the given file
   * @param path Path to a file containing a Tokenizer
   */
  static fromFile = Tokenizer.fromFile;

  /**
   * Instantiate a new Tokenizer from the given JSON string
   * @param s A JSON string representation of the Tokenizer
   */
  static fromString = Tokenizer.fromString;

  /**
   * Truncation configuration if enabled, `null` otherwise.
   *
   * @see {@link BaseTokenizer#setTruncation} to change truncation configuration
   * @see {@link BaseTokenizer#disableTruncation} to disable truncation
   * @since 0.4.0
   */
  get truncation(): Readonly<TruncationConfiguration> | null {
    return this._truncation ?? null;
  }

  /**
   * Padding configuration if enabled, `null` otherwise
   *
   * @see {@link BaseTokenizer#setPadding} to change padding configuration
   * @see {@link BaseTokenizer#disablePadding} to disable padding
   * @since 0.4.0
   */
  get padding(): Readonly<PaddingConfiguration> | null {
    return this._padding ?? null;
  }

  /**
   * Add the given tokens to the vocabulary
   *
   * @param tokens A list of tokens to add to the vocabulary.
   * Each token can either be a string, or an instance of AddedToken.
   */
  addTokens(tokens: Token[]): number {
    return this.tokenizer.addTokens(tokens);
  }

  /**
   * Add the given special tokens to the vocabulary, and treat them as special tokens.
   * The special tokens will never be processed by the model, and will be removed while decoding.
   *
   * @param tokens The list of special tokens to add.
   * Each token can either be a string, or an instance of AddedToken
   * @returns The number of tokens that were added to the vocabulary
   */
  addSpecialTokens(tokens: Token[]): number {
    return this.tokenizer.addSpecialTokens(tokens);
  }

  /**
   * Encode the given sequence
   *
   * @param sequence The sequence to encode
   * @param [pair] The optional pair sequence
   * @param [options] Some options to customize the encoding
   */
  async encode(
    sequence: InputSequence,
    pair?: InputSequence,
    options?: EncodeOptions
  ): Promise<Encoding> {
    const encode = promisify(this.tokenizer.encode.bind(this.tokenizer));
    const rawEncoding = await encode(sequence, pair ?? null, options ?? null);
    return new Encoding(rawEncoding);
  }

  /**
   * Encode the given sequences or pair of sequences
   *
   * @param sequences A list of sequences or pair of sequences.
   * The list can contain both at the same time.
   * @param [options] Sope options to customize the encoding
   */
  async encodeBatch(
    sequences: EncodeInput[],
    options?: EncodeOptions
  ): Promise<Encoding[]> {
    const encodeBatch = promisify(this.tokenizer.encodeBatch.bind(this.tokenizer));
    const rawEncodings = await encodeBatch(sequences, options);
    return rawEncodings.map((e) => new Encoding(e));
  }

  /**
   * Decode the given list of ids to a string sequence
   *
   * @param ids A list of ids to be decoded
   * @param [skipSpecialTokens=true] Whether to remove all the special tokens from the output string
   */
  decode(ids: number[], skipSpecialTokens = true): Promise<string> {
    const decode = promisify(this.tokenizer.decode.bind(this.tokenizer));
    return decode(ids, skipSpecialTokens);
  }

  /**
   * Decode the list of sequences to a list of string sequences
   *
   * @param sequences A list of sequences of ids to be decoded
   * @param [skipSpecialTokens=true] Whether to remove all the special tokens from the output strings
   */
  decodeBatch(ids: number[][], skipSpecialTokens = true): Promise<string[]> {
    const decodeBatch = promisify(this.tokenizer.decodeBatch.bind(this.tokenizer));
    return decodeBatch(ids, skipSpecialTokens);
  }

  /**
   * Enable/change truncation with specified options
   *
   * @param maxLength The maximum length at which to truncate
   * @param [options] Additional truncation options
   * @returns Full truncation configuration
   */
  setTruncation(
    maxLength: number,
    options?: TruncationOptions
  ): Readonly<TruncationConfiguration> {
    const result = this.tokenizer.setTruncation(maxLength, options);
    return (this._truncation = result);
  }

  /**
   * Disable truncation
   */
  disableTruncation(): void {
    this.tokenizer.disableTruncation();
    delete this._truncation;
  }

  /**
   * Enable/change padding with specified options
   * @param [options] Padding options
   * @returns Full padding configuration
   */
  setPadding(options?: PaddingOptions): Readonly<PaddingConfiguration> {
    const result = this.tokenizer.setPadding(options);
    return (this._padding = result);
  }

  /**
   * Disable padding
   */
  disablePadding(): void {
    this.tokenizer.disablePadding();
    delete this._padding;
  }

  /**
   * Convert the given token id to its corresponding string
   *
   * @param id The token id to convert
   * @returns The corresponding string if it exists
   */
  idToToken(id: number): string | undefined {
    return this.tokenizer.idToToken(id);
  }

  /**
   * Convert the given token to its corresponding id
   *
   * @param token The token to convert
   * @returns The corresponding id if it exists
   */
  tokenToId(token: string): number | undefined {
    return this.tokenizer.tokenToId(token);
  }

  /**
   * Apply all the post-processing steps to the given encodings.
   * The various steps are:
   * 1. Truncate according to global params (@see setTruncation)
   * 2. Apply the PostProcessor
   * 3. Pad according to global params (@see setPadding)
   * @param encoding The main Encoding to post process
   * @param [pair] An optional pair Encoding
   * @param [addSpecialTokens=true] Whether to add special tokens. Default to `true`.
   * @since 0.6.0
   */
  postProcess(encoding: Encoding, pair?: Encoding, addSpecialTokens?: boolean): Encoding {
    const rawEncoding = this.tokenizer.postProcess(
      encoding.rawEncoding,
      pair?.rawEncoding,
      addSpecialTokens
    );

    return new Encoding(rawEncoding);
  }

  /**
   * Change the post-processor to use with this Tokenizer
   * @param postProcessor New post-processor to use
   * @throws Will throw an error if any task is running
   * @throws Will throw an error if the post-processor is already used in another Tokenizer
   */
  setPostProcessor(processor: PostProcessor): void {
    return this.tokenizer.setPostProcessor(processor);
  }

  /**
   * Save the Tokenizer as JSON to the given path
   * @param path Path to the JSON file to write
   * @param [pretty=false] Whether the JSON string should be prettified
   */
  save(path: string, pretty?: boolean): void {
    return this.tokenizer.save(path, pretty);
  }

  /**
   * Get a serialized JSON version of the Tokenizer as a string
   * @param [pretty=false] Whether the JSON string should be prettified
   */
  toString(pretty?: boolean): string {
    return this.tokenizer.toString(pretty);
  }
}

/**
 * Get the string content from a token, which can be a string or AddedToken
 * @param token The token from which get the content
 */
export function getTokenContent(token: Token): string {
  return typeof token === "string" ? token : token.getContent();
}
